kyo-core/shared/src/main/scala/kyo/Channel.scala

```diff
@@ -8,334 +9,295 @@ import scala.annotation.tailrec
+        def takeFiber()(using AllowUnsafe): Fiber.Unsafe[Closed, A]
+
+        def drain()(using AllowUnsafe): Result[Closed, Seq[A]]
+        def close()(using Frame, AllowUnsafe): Maybe[Seq[A]]
+
+        def empty()(using AllowUnsafe): Result[Closed, Boolean]
+        def full()(using AllowUnsafe): Result[Closed, Boolean]
+        def closed()(using AllowUnsafe): Boolean
+
+        def safe: Channel[A] = this
+    end Unsafe
+
+    /** WARNING: Low-level API meant for integrations, libraries, and performance-sensitive code. See AllowUnsafe for more details. */
+    object Unsafe:
+        def init[A](
+            _capacity: Int,
+            access: Access = Access.MultiProducerMultiConsumer
+        )(using initFrame: Frame, allow: AllowUnsafe): Unsafe[A] =
+            new Unsafe[A]:
+                import AllowUnsafe.embrace.danger
+
+                val queue = Queue.Unsafe.init[A](_capacity, access)
+                val takes = new MpmcUnboundedXaddArrayQueue[Promise.Unsafe[Closed, A]](8)
+                val puts  = new MpmcUnboundedXaddArrayQueue[(A, Promise.Unsafe[Closed, Unit])](8)
+
+                def capacity = _capacity
+
+                def size()(using AllowUnsafe) = queue.size()
+
+                def offer(value: A)(using AllowUnsafe) =
+                    val result = queue.offer(value)
+                    if result.contains(true) then flush()
+                    result
+                end offer
+
+                def poll()(using AllowUnsafe) =
+                    val result = queue.poll()
+                    if result.exists(_.nonEmpty) then flush()
+                    result
+                end poll
+
+                def putFiber(value: A)(using AllowUnsafe): Fiber.Unsafe[Closed, Unit] =
+                    val promise = Promise.Unsafe.init[Closed, Unit]()
+                    val tuple   = (value, promise)
+                    puts.add(tuple)
+                    flush()
+                    promise
+                end putFiber
+
+                def takeFiber()(using AllowUnsafe): Fiber.Unsafe[Closed, A] =
+                    val promise = Promise.Unsafe.init[Closed, A]()
+                    takes.add(promise)
+                    flush()
+                    promise
+                end takeFiber
+
+                def drain()(using AllowUnsafe) =
+                    val result = queue.drain()
+                    if result.exists(_.nonEmpty) then flush()
+                    result
+                end drain
+
+                def close()(using frame: Frame, allow: AllowUnsafe) =
+                    queue.close().map { backlog =>
+                        flush()
+                        backlog
+                    }
+
+                def empty()(using AllowUnsafe)  = queue.empty()
+                def full()(using AllowUnsafe)   = queue.full()
+                def closed()(using AllowUnsafe) = queue.closed()
+
+                @tailrec private def flush(): Unit =
```

Comment:
the logic here is the same but using `Result` and `Maybe` since we don't need to worry about them allocating.=